src/Pact/Types/PactValue.hs

```diff
@@ -89,3 +107,216 @@ toPactValueLenient t = case toPactValue t of
+      mapLength = fromIntegral (M.size m)
+      sizeOfValues = M.foldl' (\acc pv -> acc + (sizeOf pv)) 0 m
+      sizeOfKeys = M.foldlWithKey' (\acc fk _ -> acc + (sizeOf fk)) 0 m
+
+instance (SizeOf p) => SizeOf (Guard p) where
+  sizeOf (GPact pg) = (constructorCost 1) + (sizeOf pg)
+  sizeOf (GKeySet ks) = (constructorCost 1) + (sizeOf ks)
+  sizeOf (GKeySetRef ksr) = (constructorCost 1) + (sizeOf ksr)
+  sizeOf (GModule mg) = (constructorCost 1) + (sizeOf mg)
+  sizeOf (GUser ug) = (constructorCost 1) + (sizeOf ug)
+
+instance SizeOf PactGuard where
+  sizeOf (PactGuard pid pn) =
+    (constructorCost 2) + (sizeOf pid) + (sizeOf pn)
+
+instance SizeOf KeySet where
+  sizeOf (KeySet pkArr ksPred) =
+    (constructorCost 2) + (sizeOf pkArr) + (sizeOf ksPred)
+
+instance SizeOf KeySetName where
+  -- newtype is free
+  sizeOf (KeySetName n) = sizeOf n
+
+instance SizeOf ModuleGuard where
+  sizeOf (ModuleGuard md n) =
+    (constructorCost 2) + (sizeOf md) + (sizeOf n)
+
+instance (SizeOf p) => SizeOf (UserGuard p) where
+  sizeOf (UserGuard n arr) =
+    (constructorCost 2) + (sizeOf n) + (sizeOf arr)
+
+instance SizeOf Name where
+  sizeOf (QName qn) = (constructorCost 1) + sizeOf qn
+  sizeOf (Name bn) = (constructorCost 1) + sizeOf bn
+
+instance SizeOf QualifiedName where
+  sizeOf (QualifiedName modName n i) =
+    (constructorCost 3) + (sizeOf modName) + (sizeOf n) + (sizeOf i)
+
+instance SizeOf BareName where
+  sizeOf (BareName n i) =
+    (constructorCost 2) + (sizeOf n) + (sizeOf i)
+
+instance SizeOf ModuleName where
+  sizeOf (ModuleName mn namespace) =
+    (constructorCost 2) + (sizeOf mn) + (sizeOf namespace)
+
+instance SizeOf Info where
+  sizeOf (Info may) = sizeOf may
+
+instance SizeOf Parsed where
+  sizeOf (Parsed d len) =
+    (constructorCost 2) + (sizeOf d) + (sizeOf len)
+
+instance SizeOf Delta where
+  -- Most of delta's arguments are unpacked (unboxed) Int64
+  -- Most unboxed types just take one word
+  sizeOf (Columns _ _) = 2 * wordSize
+  sizeOf (Tab _ _ _) = 3 * wordSize
+  sizeOf (Lines _ _ _ _) = 4 * wordSize
+  -- bs is the only packed argument
+  sizeOf (Directed bs _ _ _ _) =
+    (constructorCost 1) + (4 * wordSize) + (sizeOf bs)   
+
+instance SizeOf Code where
+  -- newtype is free
+  sizeOf (Code c) = sizeOf c
+
+instance (SizeOf a, SizeOf b) => SizeOf (a,b) where
+  sizeOf (a,b) = (constructorCost 3) + (sizeOf a) + (sizeOf b)
+
+instance (SizeOf a) => SizeOf (Maybe a) where
+  sizeOf (Just e) = (constructorCost 1) + (sizeOf e)
+  sizeOf Nothing = constructorCost 0
+
+instance SizeOf NamespaceName where
+  -- newtype is free
+  sizeOf (NamespaceName n) = sizeOf n
+
+instance (SizeOf a) => SizeOf [a] where
+  sizeOf arr = arrSize
+    where
+      arrSize = ((1 + (3 * arrLength)) * wordSize) + sizeOfContents
+      arrLength = fromIntegral (L.length arr)
+      sizeOfContents = L.foldl' (\acc e -> acc + (sizeOf e)) 0 arr
+
+instance SizeOf PublicKey where
```

Comment:
Some of these should be derivable once you have the underlying type implemented, ie you have `ByteString` below and I believe PublicKey is a newtype, so it should be trivally derivable. This also will move some of the instancing out of this file.

Reply:
See latest commits. Moved instances of Pact data types to their respective modules and derived newtypes.